x-pack/plugins/task_manager/server/metrics/task_metrics_collector.test.ts

```diff
@@ -0,0 +1,393 @@
+            {
+              key: 'taskType2',
+              doc_count: 1,
+              overdueByHistogram: {
+                buckets: [{ key: 0, doc_count: 1 }],
+              },
+            },
+          ],
+        },
+      },
+    });
+    mockTaskStore.aggregate.mockResolvedValueOnce({
+      took: 1,
+      timed_out: false,
+      _shards: { total: 1, successful: 1, skipped: 0, failed: 0 },
+      hits: { total: { value: 0, relation: 'eq' }, max_score: null, hits: [] },
+      aggregations: {
+        overallOverdueByHistogram: {
+          buckets: [],
+        },
+        byTaskType: {
+          doc_count_error_upper_bound: 0,
+          sum_other_doc_count: 0,
+          buckets: [],
+        },
+      },
+    });
+    mockTaskStore.aggregate.mockResolvedValueOnce({
+      took: 1,
+      timed_out: false,
+      _shards: { total: 1, successful: 1, skipped: 0, failed: 0 },
+      hits: { total: { value: 1, relation: 'eq' }, max_score: null, hits: [] },
+      aggregations: {
+        overallOverdueByHistogram: {
+          buckets: [{ key: 0, doc_count: 1 }],
+        },
+        byTaskType: {
+          doc_count_error_upper_bound: 0,
+          sum_other_doc_count: 0,
+          buckets: [
+            {
+              key: 'taskType3',
+              doc_count: 1,
+              overdueByHistogram: {
+                buckets: [{ key: 0, doc_count: 1 }],
+              },
+            },
+          ],
+        },
+      },
+    });
+    const pollInterval = 100;
+    const halfInterval = Math.floor(pollInterval / 2);
+
+    const taskManagerMetricsCollector = new TaskManagerMetricsCollector({
+      logger,
+      pollInterval,
+      store: mockTaskStore,
+    });
+    const handler = jest.fn();
+    taskManagerMetricsCollector.events.subscribe(handler);
+
+    expect(mockTaskStore.aggregate).toHaveBeenCalledTimes(1);
+    expect(mockTaskStore.aggregate).toHaveBeenCalledWith({
+      aggs: {
+        overallOverdueByHistogram: {
+          histogram: {
+            field: 'overdueBySeconds',
+            min_doc_count: 1,
+            interval: 10,
+          },
+        },
+        byTaskType: {
+          terms: { field: 'task.taskType', size: 500 },
+          aggs: {
+            overdueByHistogram: {
+              histogram: {
+                field: 'overdueBySeconds',
+                interval: 10,
+              },
+            },
+          },
+        },
+      },
+      runtime_mappings: {
+        overdueBySeconds: {
+          type: 'long',
+          script: {
+            source: `
+                def runAt = doc['task.runAt'];
+                if(!runAt.empty) {
+                  emit((new Date().getTime() - runAt.value.getMillis()) / 1000);
+                } else {
+                  def retryAt = doc['task.retryAt'];
+                  if(!retryAt.empty) {
+                    emit((new Date().getTime() - retryAt.value.getMillis()) / 1000);
+                  } else {
+                    emit(0);
```

Comment:
Very low probability but this sounds like just in time :) maybe `-1` ? don't know how it looks on the histogram though.

Reply:
We're using 10 second buckets since any task has to be slightly overdue to be claimed by task manager. This means "just in time" tasks will get grouped together with "just slightly overdue" tasks which I think is fine since we really want to threshold on a higher overdue by time to catch the scenarios where task manager isn't keeping up. Does that make sense?

Reply:
Yes right, it will be grouped with some other runs... makes sense thanks.